net: call getaddrinfo() in detachable thread to prevent stalling

pinheadmz opened the pull request:
Closes https://github.com/bitcoin/bitcoin/issues/16778
Replaces https://github.com/bitcoin/bitcoin/pull/27505

The library call `getaddrinfo()` has no internal timeout and depending on the user's system may wait indefinitely for a response when looking up a hostname in the DNS. By making that call in a separate thread, we can abandon it completely after some timeout (currently in this PR, 2 seconds).

TODO:

- [ ] We could make the polling loop interruptible but I'm not sure the best approach to that, since these functions don't live in a class with a flag like interruptNet

pinheadmz committed:
net: call getaddrinfo() in detachable thread to prevent stalling

dergoegge commented on src/test/netbase_tests.cpp:
```diff
@@ -158,6 +158,19 @@ BOOST_AUTO_TEST_CASE(embedded_test)
+BOOST_AUTO_TEST_CASE(dns_lookup)
+{
+    CNetAddr addr;
+    bool ret = LookupHost("nic.com", addr, /*fAllowLookup=*/true);
+    BOOST_CHECK(ret);
+    BOOST_CHECK(addr.IsRoutable());
```
Making actual requests from the unit tests is not great.

If you want to add a unit test, you could introduce a way to mock `getaddrinfo` or `AsyncGetAddrInfo` so that you can better test our code in all the relevant scenarios (i.e. getaddrinfo errors/blocking etc.).

dergoegge commented on src/netbase.cpp:
```diff
@@ -68,17 +79,49 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
+        // Execute the DNS lookup in another thread and check it periodically for completion.
+        // After sufficiently waiting, abandon the thread entirely. The user's system
+        // may wait forever for a DNS response that never returns.
+        std::thread thread(AsyncGetAddrInfo, &req);
+        int checks{0};
+        while (true)
+        {
+            {
+                LOCK(gai_mutex);
+                if (req.complete) break;
+            }
+            // Check every 100ms for 2 seconds
+            if (++checks > 20) break;
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+        thread.detach();
```
nit: Maybe use https://en.cppreference.com/w/cpp/thread/future/wait_until instead?

pinheadmz commented:
Still waiting for CI to finish but I removed the need for internet access by looking up the name `"localhost"` instead. That is a non-numeric lookup that `getaddrinfo()` can do without the actual DNS. I think I also fixed the ipv6 test issue.

I'm interested in following the worker pool idea to make these requests async as opposed to new threads.

dergoegge commented on src/netbase.cpp:
```diff
@@ -68,17 +79,49 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
+            if (++checks > 20) break;
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+        thread.detach();
```
Afaict this won't work here, because the thread might not finish and then the pool would just get clogged (see comment about abandoning the thread).

Performance should also not really matter here since this doesn't happen that often and startup time for a thread vs. time of a DNS lookup is probably negligible.

furszy replied:
> Afaict this won't work here, because the thread might not finish and then the pool would just get clogged (see comment about abandoning the thread).

k, `getaddrinfo` doesn't have a timeout and `getaddrinfo_a` seems to have a segfault. Cool..

> Performance should also not really matter here since this doesn't happen that often and startup time for a thread vs. time of a DNS lookup is probably negligible.

I was thinking more about context switching. Isn't this used for `OpenNetworkConnection` too? So, it would create a new thread for every connection attempt? (unless the lookups are numeric, which guess that is the rule here..).

And, following the worst case scenario, what if the user call `addnode` a good number of times? we would get a lot of detached threads? (unless we set a limit for the amount of lookups that can be resolved in parallel).

pinheadmz commented:
added call to `thread.join()` in the "happy path" so the only threads we should be abandoning are those that time out. Also setup the unit test to stub in a black hole for `getaddressinfo` to ensure the 2-second timeout is enforced. On my machine at least with thread and memory sanitizers I'm not getting any errors for the abandoned thread. I'm going to try some more experiments with a mainnet node as well.

pinheadmz commented:
...got thread sanitizer errors running locally on main when my local DNS is black-holed. Fixed that with a shared pointer.

pinheadmz commented on src/netbase.cpp:
```diff
@@ -40,7 +40,9 @@ bool fNameLookup = DEFAULT_NAME_LOOKUP;
 static std::atomic<bool> interruptSocks5Recv(false);
 
-std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_lookup)
+static GlobalMutex gai_mutex;
```
TIL I can probably replace this with `std::atomic_bool`

pinheadmz commented:
> Needs rebase?

ty. 🤞

luke-jr commented on src/netbase.cpp:
```diff
@@ -54,31 +54,68 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
+    if (allow_lookup) {
+        // Execute the DNS lookup in another thread and check it periodically for completion.
+        // After sufficiently waiting, abandon the thread entirely. The user's system
+        // may wait forever for a DNS response that never returns.
```
Probably not good to ignore it never returning... it's leaking resources. Maybe kill it and/or tell the user somehow? :/

pinheadmz replied:
I added a log message but not sure what else to do. The problem is that `getaddrinfo()` is blocking and `getaddrinfo_a()` segfaults 😭 As long as the DNS request resolves *eventually* the thread will close, even if bitcoind has given up and detached the thread. If `getaddrinfo()` hangs forever then yes you're right that is a leak, but the user also has a serious networking issue.

luke-jr commented on src/test/netbase_tests.cpp:
```diff
@@ -156,6 +156,42 @@ BOOST_AUTO_TEST_CASE(embedded_test)
+{
+    // Non-numeric lookup that does not require actual DNS service
+    BOOST_CHECK(LookupHost("localhost", /*fAllowLookup=*/true).has_value());
+    BOOST_CHECK(!LookupHost("fail.doesnotexist", /*fAllowLookup=*/true).has_value());
```
Won't this still make an externally-visible DNS request?

pinheadmz replied:
Yep you are absoutley right, I'll remove.
